SAMZA-2127: Upgrade to Kafka 2.0

mynameborat opened the pull request:
Initial PR for Kafka 2.0 upgrade

cameronlee314 commented on samza-test/src/test/scala/org/apache/samza/test/integration/StreamTaskTestUtil.scala:
```diff
@@ -257,23 +250,27 @@ class StreamTaskTestUtil {
+        } else {
+          messages += new String(message.value, "UTF-8")
+        }
+        System.err.println("StreamTaskTestUtil.readAll(): offset=%s, message=%s" format (message.offset, messages.last))
```
I know this was already here, but it doesn't seem like an error. Could it just be removed, or at least changed to System.out?

mynameborat commented on samza-kafka/src/test/scala/org/apache/samza/system/kafka/TestKafkaSystemAdmin.scala:
```diff
@@ -257,17 +211,22 @@ class TestKafkaSystemAdmin {
+    val records = consumer.poll(Duration.ofMillis(1000))
+    consumer.close()
+
+    assertTrue(records.count() > 2)
```
Forgot to push the local change.
Fixed it.

mynameborat commented on samza-kafka/src/test/scala/org/apache/samza/checkpoint/kafka/TestKafkaCheckpointManager.scala:
```diff
@@ -110,15 +110,12 @@ class TestKafkaCheckpointManager extends KafkaServerTestHarness {
     // check that start actually creates the topic with log compaction enabled
-    val zkClient = ZkUtils(zkConnect, 6000, 6000, JaasUtils.isZkSecurityEnabled())
-    val topicConfig = AdminUtils.fetchEntityConfig(zkClient, ConfigType.Topic, checkpointTopic)
+    val topicConfig = adminZkClient.getAllTopicConfigs().getOrElse(checkpointTopic, new Properties())
```
`adminZkClient` is a very light weight utility wrapper around `zkClient` and doesn't have any lifecycle methods. However, `ZookeeperTestHarness` parent of `KafkaServerTestHarness` takes care of managing the lifecycle of zkClient.

mynameborat commented on samza-kafka/src/test/scala/org/apache/samza/system/kafka/TestKafkaSystemAdmin.scala:
```diff
@@ -301,12 +260,6 @@ class TestKafkaSystemAdmin {
-    val topicMetadata = topicMetadataMap(topic)
-    val partitionMetadata = topicMetadata.partitionsMetadata.head
-    assertEquals(0, partitionMetadata.partitionId)
-    assertEquals(3, partitionMetadata.replicas.size)
```
Yeah. I tried to keep this by using APIs from `adminClient` and `adminZkClient`but in vain. I'd have to replicate the logic in the file `ClientsUtilTopicMetadataStore.scala` that was removed as part of the PR. Since we no longer have any active code path within our codebase that reads this data, I chose to remove it.

cameronlee314 replied:
The end-to-end check seems unnecessary anyways for a unit test.
However, since you are removing that check, it might be good to have a test which uses Mockito to verify that the correct configurations are passed to the admin client when creating topics. If I recall correctly, in the past, we may have had some issues which involved replication factors (not sure if they were this exact flow though).

mynameborat replied:
Agreed. Let me do the Mockito way 👍

mynameborat commented:
Summary of 44d8c97
* Combined StandaloneIntegrationTestHarness and AbstractIntegrationTestHarness into IntegrationTestHarness. 
* Modified some more tests to use `adminClient` APIs and got rid of deprecated `zkUtils`
* Addressed Cameron's review comments.
* Fixed a bug around metric where in records-lag is not longer prefixed with the topic partition rather topic partitions are part of tags within the metric names.

cameronlee314 commented on samza-kafka/src/test/scala/org/apache/samza/system/kafka/TestKafkaSystemAdmin.scala:
```diff
@@ -89,49 +91,49 @@ object TestKafkaSystemAdmin extends KafkaServerTestHarness {
+        val topicDescription = topicDescriptionFutures.get(500, TimeUnit.MILLISECONDS)
+            .get(topic)
 
         done = expectedPartitionCount == topicDescription.partitions().size()
```
1. Can `topicDescription` be null?
2. If done is true, then can we break out of this loop?

mynameborat replied:
1. Nope. 
2. Thanks for bringing it up. I found a corner case with current implementation. I changed it a little without introducing additional control flow statements.

cameronlee314 commented on samza-test/src/test/java/org/apache/samza/test/framework/StreamApplicationIntegrationTestHarness.java:
```diff
@@ -295,4 +227,20 @@ public Config getConfig() {
   }
+
+  @Override
+  protected KafkaProducer createProducer() {
```
It seems a bit odd that this test harness needs slightly different serialization/deserialization for createProducer/createConsumer compared to `IntegrationTestHarness`. If it is difficult to make them consistent, could you please put a comment about why you have these overrides?
You could also consider having helpers in `IntegrationTestHarness` to build the configs, just in case someone needs to add something in the future.

mynameborat replied:
I added a comment and wired in the serde through configuration.

cameronlee314 commented on samza-test/src/test/java/org/apache/samza/test/processor/TestZkLocalApplicationRunner.java:
```diff
@@ -159,42 +152,25 @@ public void setUp() {
+            .map(topic -> new NewTopic(topic, 5, (short) 1))
+            .collect(Collectors.toList());
+
+    createTopics(newTopics);
```
Should this fail if `createTopics` returns false?

cameronlee314 commented on samza-test/src/test/scala/org/apache/samza/test/integration/StreamTaskTestUtil.scala:
```diff
@@ -133,43 +130,36 @@ object StreamTaskTestUtil {
+    adminClient.createTopics(Collections.singleton(new NewTopic(INPUT_TOPIC, TOTAL_TASK_NAMES, REPLICATION_FACTOR.shortValue())))
   }
 
   def validateTopics {
```
Not sure if this is feasible/practical to combine, but it looks like some of this code is similar to code from `TestKafkaSystemAdmin`, such as `validateTopics`.

mynameborat replied:
It is doable and I considered it. I deferred it due to following

1. I didn't want to add more churn to scala code. StreamTaskTestUtil still uses some of the TestUtils deprecated code and I chose to clean up all at once as a separate PR (maybe rewrite them in java)
2. I didn't want to expand the scope of this PR.

I have created a follow up ticket (https://issues.apache.org/jira/browse/SAMZA-2145) to clean up the test and consolidate the common code to some helper or move it to test harness.

shanthoosh commented on gradle/dependency-versions.gradle:
```diff
@@ -34,7 +34,7 @@
   joptSimpleVersion = "3.2"
   junitVersion = "4.12"
-  kafkaVersion = "0.11.0.2"
+  kafkaVersion = "2.0.1"
```
After this, through kafka 2.0, we would pull in slf4j-log4j12 transitively. 

Do we know if this would cause logging problems with different log4j-1/log4j-2 combinations that exists in samza.  Or it doesn't matter?

mynameborat replied:
We already pull in `slf4j-log4j12` as part of build time dependency in samza-log4j module. Since we have a direct pinned dependency, I suppose it shouldn't matter.

Let me keep an eye for any potential logging issues.

shanthoosh commented on samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemAdmin.java:
```diff
@@ -668,22 +631,86 @@ protected Properties createAdminClientProperties() {
+          .collect(Collectors.toMap(Function.identity(), streamId -> {
+            Properties properties = new Properties();
+            properties.putAll(streamConfig.getStreamProperties(streamId));
+            properties.putIfAbsent("retention.ms", String.valueOf(KafkaConfig.DEFAULT_RETENTION_MS_FOR_BATCH()));
```
Would be better to switch from hard-coded string: `retention.ms` to `TopicConfig.RETENTION_MS_CONFIG ` here.

shanthoosh commented on samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemAdmin.java:
```diff
@@ -668,22 +631,86 @@ protected Properties createAdminClientProperties() {
+
+  private Properties getCoordinatorStreamProperties(KafkaConfig config) {
+    Properties coordinatorStreamProperties = new Properties();
+    coordinatorStreamProperties.put("cleanup.policy", "compact");
```
1. Better to replace hard-coded string(`cleanup.policy`) with `TopicConfig.RETENTION_MS_CONFIG `. 
2. Better to replace hard-coded string: `compact` with `TopicConfig.CLEANUP_POLICY_COMPACT `.
3. Better to replace `segment.policy` with `TopicConfig. SEGMENT_BYTES_CONFIG `

mynameborat commented on samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemAdmin.java:
```diff
@@ -628,10 +589,12 @@ public void validateStream(StreamSpec streamSpec) throws StreamValidationExcepti
+          .collect(Collectors.toMap(entry ->
+              new TopicPartition(entry.getKey().getStream(), entry.getKey().getPartition().getPartitionId()),
+              entry -> RecordsToDelete.beforeOffset(Long.parseLong(entry.getValue()) + 1)));
+      adminClient.deleteRecords(recordsToDelete);
```
Done 1 & 2 
3. We have been using AdminClient for all flows except deletion path. We don't have any of the backoff configurations configurable for the other flows and I plan to keep it the same for deletion too.

We can follow up if there is a need for it in a separate PR

mynameborat commented on samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemAdmin.java:
```diff
@@ -628,10 +589,12 @@ public void validateStream(StreamSpec streamSpec) throws StreamValidationExcepti
+          .collect(Collectors.toMap(entry ->
+              new TopicPartition(entry.getKey().getStream(), entry.getKey().getPartition().getPartitionId()),
+              entry -> RecordsToDelete.beforeOffset(Long.parseLong(entry.getValue()) + 1)));
+      adminClient.deleteRecords(recordsToDelete);
```
Good point. I am caught up on this since the existing behavior wasn't blocking either and does a best effort on the delete. I debated between retaining existing behavior vs blocking. Since blocking on this would mean blocking commit thread which does seem significant especially given this is an optimization to reduce the data footprint for our intermediate topics.

Let me log an error in case of failures on this future.

cameronlee314 commented on samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemAdmin.java:
```diff
@@ -594,7 +593,13 @@ public void deleteMessages(Map<SystemStreamPartition, String> offsets) {
               entry -> RecordsToDelete.beforeOffset(Long.parseLong(entry.getValue()) + 1)));
-      adminClient.deleteRecords(recordsToDelete);
+
+      adminClient.deleteRecords(recordsToDelete).all().whenComplete((ignored, exception) -> {
```
Just to double check, `whenComplete` is async, right?

mynameborat replied:
Yes

cameronlee314 commented on samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemAdmin.java:
```diff
@@ -594,7 +593,13 @@ public void deleteMessages(Map<SystemStreamPartition, String> offsets) {
+
+      adminClient.deleteRecords(recordsToDelete).all().whenComplete((ignored, exception) -> {
+        if (exception != null) {
+          LOG.error("Delete message failed for SSPs {} due to {}", offsets.keySet(), exception);
```
I think this will end up not printing out the stack trace. I think you need to use the `LOG.error(String, Exception)` instead.

mynameborat replied:
1. Since this is an optimization in the background to clean up data for intermediate topics, it is not be desirable to bubble it up and kill the application.
2. The return future is of type Void and ignored doesn't carry any information.

cameronlee314 commented on samza-test/src/test/java/org/apache/samza/test/framework/StreamApplicationIntegrationTestHarness.java:
```diff
@@ -229,18 +232,32 @@ public Config getConfig() {
+     * The default serde for the test producer in IntegrationTestHarness uses
+     * byte serde for value and string for key.
+     */
+    producerProps.setProperty(KEY_SERIALIZER_CLASS_CONFIG, STRING_SERIALIZER);
```
Could you implement this by calling `super.createProducerConfigs`, and then just override the serializers? Same for `createConsumerConfigs`.

shanthoosh commented on gradle/dependency-versions.gradle:
```diff
@@ -34,7 +34,7 @@
   joptSimpleVersion = "3.2"
   junitVersion = "4.12"
-  kafkaVersion = "0.11.0.2"
+  kafkaVersion = "2.0.1"
```
Discussed offline.

Since we're bumping up kafka-client version to 2.0, it would be better to run the  perf tests. If possible try running [kafka-read-write-perf](https://github.com/apache/samza/blob/master/samza-test/src/main/config/perf/kafka-read-write-performance.properties) test before landing this patch and verify there's no performance degradation(in terms of QPS/poll-latency).

mynameborat replied:
will do

mynameborat commented on samza-kafka/src/test/scala/org/apache/samza/system/kafka/TestKafkaSystemAdmin.scala:
```diff
@@ -169,59 +165,21 @@ class TestKafkaSystemAdmin {
-    // Create an empty topic with 50 partitions, but with no offsets.
-    createTopic(TOPIC, 50)
-    validateTopic(TOPIC, 50)
+    // Reduced the partition count to 25 instead of 50 to prevent tests from failing. In the current setup,
```
Discussed offline. The essence of the tests remains the same except the higher number of partitions doesn't play well with the current Embedded Kafka setup (3 brokers) in terms of replication assignment especially with the new create APIs which used in the tests which use RackAwareMode as Enforced by default.

I didn't want to spend too much time on getting it to work since the essence of the tests still remains the same.

mynameborat commented on samza-test/src/test/java/org/apache/samza/test/harness/IntegrationTestHarness.java:
```diff
@@ -0,0 +1,192 @@
+    boolean createStatus = true;
+
+    try {
+      CreateTopicsResult resultFuture =
```
Synced up offline. 
Taking up dependency on the Samza code path for setting up environment (in this case topic creation) is unnecessary and doesn't buy in terms of code reuse.

Also, createStreams API in KafkaSystemAdmin also has bunch of boiler plate code and would mean the creation path is constrained by the parameters that are used to create SystemAdmin which compromises flexibility.

shanthoosh commented:
@mynameborat 
Thanks for the changes. Merged it to trunk.
